_posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md

```diff
@@ -0,0 +1,467 @@
+feel free to skip this part.
+
+#### A typical Metal3 Workflow
+
+The following diagram explains a typical Metal3 workflow. We will, then, go into
+details of every component.
+
+![Metal3 Typical
+Workflow](/assets/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node/metal3-typical-workflow.jpg)
+
+#### Cluster API (CAPI)
+
+CAPI is a project that simplifies the deployment and management of Kubernetes
+clusters. It provides a consistent way to create, update, and scale clusters
+through Kubernetes-native APIs. The magic of CAPI is that it abstracts away many
+of the underlying details so that you can manage clusters on different platforms
+(cloud, bare metal, etc.) in a unified way.
+
+#### Cluster API Provider Metal3 (CAPM3)
+
+CAPM3 extends CAPI to work specifically with Metal3 environments. It connects
+the dots between CAPI, BMO, and Ironic, allowing Kubernetes clusters to be
+deployed on bare-metal infrastructure. It handles tasks like provisioning new
+nodes, registering them with Kubernetes, and scaling clusters.
+
+#### Bare Metal Operator (BMO)
+
+BMO is a controller that runs inside a Kubernetes cluster and works alongside
+Ironic to manage bare-metal infrastructure. It automates the lifecycle of
+bare-metal hosts, managing things like registering new hosts, powering them on
+or off, and monitoring their status.
+
+##### Bare Metal Host (BMH)
+
+A BMH is the Kubernetes representation of a bare-metal node. It contains
+information about how to reach the node it represents, and BMO monitors its
+desired state closely. When BMO notices that a BMH object state is requested to
+change (either by a human user or CAPM3), it will decide what needs to be done
+and tell Ironic.
+
+#### Ironic & Ironic Python Agent (IPA)
+
+- Ironic is a bare-metal provisioning tool that handles tasks like booting
+servers, deploying bootable media (e.g., operating systems) to disk, and
+configuring hardware. Think of Ironic as the piece of software that manages
+actual physical servers. In a Metal3 workflow, Ironic receives orders from BMO
+and translates them into actionable steps. Ironic has multiple ways to interact
+with the machines, and one of them is the so-called " agent-based direct deploy"
+method, which is commonly used by BMO. The agent mentioned is called __Ironic
+Python Agent__ (IPA), which is a piece of software that runs on each bare-metal
+node and carries out Ironic's instructions. It interacts with the hardware
+directly, like wiping disks, configuring networks, and handling boot processes.
+
+In a typical Metal3 workflow, BMO reads the desired state of the node from the
+BMH object, translates the Kubernetes reconciling logic to concrete actions, and
+forwards them to Ironic, which, as part of the provisioning process, tells IPA
+the exact steps it needs to perform to get the nodes to desired states. During
+the first boot after node image installation,
+[cloud-init](https://cloud-init.io/)) will install Kubernetes components on the
```

Comment:
```suggestion
[cloud-init](https://cloud-init.io/) will install Kubernetes components on the
```
I would personally avoid putting so much details here because many of them may not always be correct. Cloud-init *can* be used, but so can ignition, cloud-init *may* install kubernetes components, or they may be already in the image, etc...